### PR TITLE
Alt navigation

### PR DESCRIPTION
**Pull request by RensDofferhoff**
Implementation of the ALT navigation functionality

### CODE REVIEW

File: Desktop/gui/ALTNavigation/altnavcontrol.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+{
+	if(!instance)
+		instance = new ALTNavControl();
+	return instance;
+}
+
+ALTNavControl::ALTNavControl(QObject *parent) : QObject{parent}
+{
+	qApp->installEventFilter(this);
+	currentRoot = currentNode = defaultRoot = new ALTNavScope(this);
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavcontrol.cpp]: :p

File: Desktop/gui/ALTNavigation/altnavcontrol.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	static ALTNavControl* getInstance();
+	ALTNavControl(ALTNavControl& other) = delete;
+	void operator=(const ALTNavControl&) = delete;
+
+signals:
+	void altNavInputChanged();
+	void altNavEnabledChanged();
+
+private:
+	static ALTNavControl* instance;
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavcontrol.h]: Could you rename the class variables to start with a `_` ?
It really helps in making clear where variables are coming from when looking at the code

Comment (JorisGoosen) [reply Desktop/gui/ALTNavigation/altnavcontrol.h]: With the inbuilt renaming refactor option in qtcreator it isnt such a pain luckily

Comment (JorisGoosen) [reply Desktop/gui/ALTNavigation/altnavcontrol.h]: In structs we leave the `_` out when they are public but in a class like this I think `_` is more appropriate

File: Desktop/gui/ALTNavigation/altnavigation.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ *		delegate: SomeItem
+ *		{
+ *			ALTNavigation.enabled:		true
+ *			ALTNavigation.index:		index
+ *			ALTNavigation.onTagMatch:	{ open(); }
+ *		}
+ *  }
+ *
+ *  @author Rens Dofferhoff
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavigation.h:71]: @boutinb it looks like we will need to "up" our documentation game!

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		case PRIORITY:
+			return new PriorityStrategy();
+		case INDEXED:
+			return new IndexedStrategy();
+		case PASSTHROUGH:
+			return new PassthroughStrategy();
+		case UNKNOWN:
+			return nullptr;
+		default:
+			return nullptr;
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]:
```suggestion
		case UNKNOWN:
			[[fallthrough]];
		default:
			return nullptr;
```

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			assignments[request[0].toLatin1() - 'A'] = scope;
+			assigned.insert(request[0].toLatin1());
+			scope->setPrefix(prefix + scope->getRequestedPostfix());
+		}
+		else
+			unassigned.push_back(scope);
+	}
+
+	//assign the leftover postfixes
+	std::string preferedSacrifices = "";
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]:
```suggestion
	std::string preferredSacrifices = "";
```

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	for (char option : capitalLetters)
+	{
+		if(unassigned.empty())
+			return;
+		else if (assigned.contains(option))
+			continue;
+		else
+		{
+			preferedSacrifices += option;
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]:
```suggestion
			preferredSacrifices += option;
```

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+
+	//We assigned 26 postfixes so now we need to get creative and we start assigning postfixes of lenght >= 2, like ZA,ZB...,ZZ
+	QString layerPrefix = "";
+	QString nextLayerPrefix = "";
+	//no spots left so we need a sacrifice. If no candidates is available we use Z. Who uses the letter Z anyways
+	if (assigned.size() == 26 && unassigned.size() > 0) //26 letters in alphabet
+	{
+		char sacrifice = preferedSacrifices.length() == 0 ? 'Z' : preferedSacrifices.back();
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]:
```suggestion
		char sacrifice = preferredSacrifices.length() == 0 ? 'Z' : preferredSacrifices.back();
```

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+				unassigned.pop_back();
+			}
+		}
+		layerPrefix = nextLayerPrefix;
+	}
+}
+
+void IndexedStrategy::assignPostfixes(const QObjectList &children, QString prefix)
+{
+	int maxLeadingZeros = std::log10((double)children.size()); //if needed I can write a fast integer version that never fails but I doubt it will really matter for reasonable input sizes
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp:112]: 👍

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	for (auto child : children)
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]:
```suggestion
	for (auto * child : children)
```
I know its in there but I like being able to see its a pointer ^^

File: Desktop/gui/ALTNavigation/altnavscope.cpp [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+			return;
+		}
+		//partial match
+		bool partialMatch = input.length() < scope->_prefix.length() && scope->_prefix.first(input.length()) == input;
+		if (!partialMatch)
+			scope->setScopeActive(false);
+		matchPossible |= partialMatch;
+	}
+
+	//end of the road so we disable the mode and return to root
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavscope.cpp]:
```suggestion
	//end of the road so we disable the node and return to root
```

File: Desktop/gui/ALTNavigation/altnavscope.h [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	 */
+	void registerWithParent();
+
+
+protected:
+	bool scopeActive = false;
+	QString _prefix = "";
+
+private:
+	//ALTNavigation Interface
```
Comment (JorisGoosen) [inline Desktop/gui/ALTNavigation/altnavscope.h:130]: `_` here as well

File: Desktop/mainwindow.cpp [hunk 12]
```diff
@@ -218,8 +219,9 @@ MainWindow::~MainWindow()
 	{
 		//Clean up all QML to get rid of warnings and hopefully fix https://github.com/jasp-stats/jasp-issues/issues/667
 		//Going backwards to make sure the theme isnt deleted before everything that depends on it
-		for(int i=_qml->rootObjects().size() - 1; i >= 0; i--)
+		for(int i=_qml->rootObjects().size() - 1; i >= 0; i--) {
 			delete _qml->rootObjects().at(i);
+		}
```
Comment (JorisGoosen) [inline Desktop/mainwindow.cpp]:
```suggestion
		for(int i=_qml->rootObjects().size() - 1; i >= 0; i--)
			delete _qml->rootObjects().at(i);

```

Or why add the braces?

File: QMLComponents/components/JASP/Theme/Theme.qml [hunk 13]
```diff
@@ -116,6 +121,8 @@ JaspTheme
 	itemSelectedColor:					blueDarker
 	itemSelectedNoFocusColor:			grayLighter
 
+	altNavTagColor:						"#565756"
```
Comment (JorisGoosen) [inline QMLComponents/components/JASP/Theme/Theme.qml:124]: Maybe specify one in DarkTheme.qml as well?

File: QMLComponents/jasptheme.h [hunk 14]
```diff
@@ -714,7 +725,8 @@ private slots:
 						_fontRCode,
 						_fontCode,
 						_fontGroupTitle,
-						_fontPrefOptionsGroupTitle;
+						_fontPrefOptionsGroupTitle,
+						_fontALTNavTag;
```
Comment (JorisGoosen) [inline QMLComponents/jasptheme.h:729]: Why add an extra font for this?

Not saying it is a bad idea, just curious why

**Review by JorisGoosen — CHANGES_REQUESTED** [review]
Pretty amazing work Rens!

I have some small nitpicks but overall this was a pleasure to read and it seems to work quite nicely too!

File: QMLComponents/jasptheme.h (hunk 14 above)
Comment (RensDofferhoff) [reply QMLComponents/jasptheme.h:729]: The others were a bit large. The size of the tag is determined by the font size of the text

File: Desktop/gui/ALTNavigation/altnavigation.h (hunk 3 above)
Comment (boutinb) [reply Desktop/gui/ALTNavigation/altnavigation.h:71]: Yes I'm impressed!

**Review by boutinb — APPROVED** [review]
I have tested it, and it works really well. Of course some area still misses for the key navigation (variables types/labels, Computed Columns, Filters), but that is already so much better than what it was.

File: Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp (hunk 7 above)
Comment (RensDofferhoff) [reply Desktop/gui/ALTNavigation/altnavpostfixassignmentstrategy.cpp]: oops ;)

File: Desktop/gui/ALTNavigation/altnavcontrol.cpp (hunk 1 above)
Comment (RensDofferhoff) [reply Desktop/gui/ALTNavigation/altnavcontrol.cpp]: > qApp->installEventFilter(this);

 I will consider another method in the future.
It does slow down the entire event loop a little.
There is no obvious method though

Comment (JorisGoosen) [reply Desktop/gui/ALTNavigation/altnavcontrol.cpp]: I think this is fine, the slowdown should be tiny.

**Review by JorisGoosen — APPROVED** [review]

**Comment by RensDofferhoff** [comment]
I have implemented @JorisGoosen 's feedback.

> I have tested it, and it works really well. Of course some area still misses for the key navigation (variables types/labels, Computed Columns, Filters), but that is already so much better than what it was.

Its the same for the results window portion.
Most analysis and menu interactions can now be done by keyboard. 
There is a lot/near endless amount of work that could  be done still.
We can discuss internally what improvements we pursue in future.
If you have ideas for useful tag additions let me know!

**Pull request merged**